### DISCUSSION TITLE
DNS resolution in MacOS with Nodejs S3-Client

### DISCUSSION THREAD

**Discussion by gdonoso94** [aws/aws-sdk-js-v3#3877] [Q&A]
Hi all,

I've been struggling with this for a while and I'm not sure if this is a bug or not, so I preferred to open a new discussion about it.

**Before all: when I run this in a Linux environment (Ubuntu) it works fine**, as expected.

I'm trying to test the integration of an S3 upload using testcontainers and Localstack. For that I raise the Localstack Docker container and I create a bucket named `test`, so far so good. I'm not sure whether the problem comes when I try to create the S3-Client using `@aws-sdk/client-s3` or when I upload the file using the `Upload` command from. `@aws-sdk/lib-storage`; turns out it and it's unable to resolve `test.localhost`. This is the output:

```
Error: getaddrinfo ENOTFOUND test.localhost
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:84:26) {
  errno: -3008,
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'test.localhost',
  '$metadata': { attempts: 1, totalRetryDelay: 0 }
}
```

I've been reading around and looks like it has something to do with using dns.resolve() or dns.lookup() but I haven't come up with a solution yet.

Can somebody help me out here?

Thanks in advance.

**Comment by RanVaknin** [comment]
Hi @gdonoso94,

Do you mind including your actual code? 

Thanks.
Ran

**Reply by gdonoso94** [reply]
Hi @RanVaknin, sorry for the delay in coming back.

I basically have this s3Upload factory:

```export const s3Upload: S3UploadFactoryFunction =
  ({ client }) =>
  ({ file, path, fileName, mimeType }) => {
    const params = {
      Bucket: S3_BUCKET,
      Key: `${path}/${fileName}`,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      Body: Readable.from(file),
      ContentType: mimeType,
    };

    return new Upload({ client, params });
  };
```

That returns the Upload object from `@aws-sdk/lib-storage`.

In the test I spin up the Localstack container like this:

```
beforeAll(async () => {
    container = await new GenericContainer("localstack/localstack").withExposedPorts(4566).start();
    client = initializeS3Client({
      endpoint: `http://${container.getHost()}:${container.getMappedPort(4566)}`,
      accessKeyId: S3_ACCESS_KEY_ID,
      secretAccessKey: S3_SECRET_ACCESS_KEY,
      region: S3_REGION,
    });
    await client.send(new CreateBucketCommand({ Bucket: "test" }));
  });
```

Where initializeS3Client is basically a factory for S3Client from `@aws-sdk/client-s3`.

Looks like the problem arises in `client.send(new CreateBucketCommand({ Bucket: "test" }))`.  The bucket cannot be created. As far as I researched and mentioned in the previous message it has something to do with how node resolves the DNS, using `dns.resolve` or `dns.lookup`...

I'm using node 18, btw.

Thanks for your help

**Comment by gdonoso94** [comment]
Copyin and pasting here. I don't know if this will update the label. Sorry if this is not the procedure:

Hi @RanVaknin, sorry for the delay in coming back.

I basically have this s3Upload factory:

  ({ client }) =>
  ({ file, path, fileName, mimeType }) => {
    const params = {
      Bucket: S3_BUCKET,
      Key: `${path}/${fileName}`,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      Body: Readable.from(file),
      ContentType: mimeType,
    };

    return new Upload({ client, params });
  };
That returns the Upload object from @aws-sdk/lib-storage.

In the test I spin up the Localstack container like this:

beforeAll(async () => {
    container = await new GenericContainer("localstack/localstack").withExposedPorts(4566).start();
    client = initializeS3Client({
      endpoint: `http://${container.getHost()}:${container.getMappedPort(4566)}`,
      accessKeyId: S3_ACCESS_KEY_ID,
      secretAccessKey: S3_SECRET_ACCESS_KEY,
      region: S3_REGION,
    });
    await client.send(new CreateBucketCommand({ Bucket: "test" }));
  });
Where initializeS3Client is basically a factory for S3Client from @aws-sdk/client-s3.

Looks like the problem arises in client.send(new CreateBucketCommand({ Bucket: "test" })). The bucket cannot be created. As far as I researched and mentioned in the previous message it has something to do with how node resolves the DNS, using dns.resolve or dns.lookup...

I'm using node 18, btw.

Thanks for your help

**Comment by egimenos** [comment]
Just in case someone have the same problem, it gets fixed by setting the `forcePathStyle: true` at S3Client config object.

Reference: https://github.com/ar90n/serverless-s3-local/issues/492#issuecomment-1058519846
